modules/currency-exchange/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+pub trait Trait: frame_system::Trait {
+	/// Handler for transaction submission result.
+	type OnTransactionSubmitted: OnTransactionSubmitted<Self::AccountId>;
+	/// Peer blockchain type.
+	type PeerBlockchain: Blockchain;
+	/// Peer blockchain transaction parser.
+	type PeerMaybeLockFundsTransaction: MaybeLockFundsTransaction<
+		Transaction = <Self::PeerBlockchain as Blockchain>::Transaction,
+	>;
+	/// Map between blockchains recipients.
+	type RecipientsMap: RecipientsMap<
+		PeerRecipient = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Recipient,
+		Recipient = Self::AccountId,
+	>;
+	/// This blockchain currency amount type.
+	type Amount;
+	/// Converter from peer blockchain currency type into current blockchain currency type.
+	type CurrencyConverter: CurrencyConverter<
+		SourceAmount = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Amount,
+		TargetAmount = Self::Amount,
+	>;
+	/// Something that could grant money.
+	type DepositInto: DepositInto<Recipient = Self::AccountId, Amount = Self::Amount>;
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Invalid peer blockchain transaction provided.
+		InvalidTransaction,
+		/// Peer transaction has invalid amount.
+		InvalidAmount,
+		/// Peer transaction has invalid recipient.
+		InvalidRecipient,
+		/// Cannot map from peer recipient to this blockchain recipient.
+		FailedToMapRecipients,
+		/// Failed to convert from peer blockchain currency to this blockhain currency.
+		FailedToConvertCurrency,
+		/// Deposit has failed.
+		DepositFailed,
+		/// Transaction is not finalized.
+		UnfinalizedTransaction,
+		/// Transaction funds are already claimed.
+		AlreadyClaimed,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Imports lock fund transaction of the peer blockchain.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn import_peer_transaction(
+			origin,
+			proof: <<T as Trait>::PeerBlockchain as Blockchain>::TransactionInclusionProof,
+		) -> DispatchResult {
+			let submitter = frame_system::ensure_signed(origin)?;
+
+			// ensure that transaction is included in finalized block that we know of
+			//
+			// note that the transaction itself is actually redundant here, because it
+			// will probably be a part of proof itself (like it is now), or could be
+			// reconstructed from proof during verification
+			//
+			// leaving it here for now just for simplicity
+			let transaction = <T as Trait>::PeerBlockchain::verify_transaction_inclusion_proof(
+				&proof,
+			).ok_or_else(|| Error::<T>::UnfinalizedTransaction)?;
+
+			// parse transaction
+			let transaction = <T as Trait>::PeerMaybeLockFundsTransaction::parse(&transaction)
+				.map_err(Error::<T>::from)?;
+			let transfer_id = transaction.id;
+			ensure!(
+				!Transfers::<T>::contains_key(&transfer_id),
+				Error::<T>::AlreadyClaimed
+			);
```

Comment:
The exclamation mark is easy to miss, since we need to be extra careful in the runtime I suggest the following

```suggestion
                      let has_not_been_claimed = !Transfers::<T>::contains_key(&transfer_id);
			ensure!(
				has_not_been_claimed,
				Error::<T>::AlreadyClaimed
			);
```